### ISSUE TITLE
The function argument of recursive data type raises Argument of type error.

### ISSUE THREAD

**Issue by yassun7010** [microsoft/pyright#7262]
**Describe the bug**
The function argument of recursive data type raises Argument of type error.

**Code or Screenshots**
```python
from typing import TypeAlias

JsonValue: TypeAlias = (
    str | int | float | bool | None | dict[str, "JsonValue"] | list["JsonValue"]
)


def method(data: JsonValue) -> None:
    pass


method({"a": {"b": 2}})  # Pass

data = {"a": {"b": 2}}

method(data)  # Falied
# Argument of type "dict[str, dict[str, int]]" cannot be assigned to parameter "data" of type "JsonValue" in function "method"

```

**VS Code extension or command-line**

pyright 1.1.350

**Comment by erictraut** [comment]
Pyright is working correctly here, so this isn't a bug.

The problem in your code is that the `data` variable has no type declaration, so its type is determined by normal dictionary inference rules. In this case, its type is inferred as `dict[str, dict[str, int]]`. This type is not compatible with `JsonValue` because the type parameters in `dict` are invariant and must match exactly. 

If you want this to type check without errors, you will need to declare the type of `data` so pyright uses bidirectional type inference to infer the type of the assigned expression as `JsonValue`.

```python
data: JsonValue = {"a": {"b": 2}}
```
